### PR TITLE
Easy-2033 check status before deposit updates

### PR DESCRIPTION
**Pull request by stephanHogenboom**
Fixes EASY-2033

- [x] test on deasy
- [x] update swagger

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

**Comment by stephanHogenboom** [comment]
Renamed branch

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/StateInfo.scala [hunk 1]
```diff
@@ -23,14 +23,20 @@ import scala.collection.Seq
 import scala.util.{ Failure, Success, Try }
 
 case class StateInfo(state: State, stateDescription: String) {
-  def isDeletable: Try[Unit] = {
+  def canDelete: Try[Unit] = {
     if (StateInfo.deletableStates.contains(state)) Success(())
     else Failure(new IllegalStateException(s"Deposit has state $state, can only delete deposits with one of the states: ${ StateInfo.deletableStates.mkString(", ") }"))
   }
+
+  def canUpdate: Try[Unit] = {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/docs/StateInfo.scala]: 1. `canUpdate` and `canDelete` look awfully similar. If you have a third `canXXX`, I would suggest refactoring them a bit.
2. These error messages are very descriptive. Is there any chance they end up in {{easy-deposit-ui}}. @jo-pol could you take these into account when you look at EASY-2016?

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 2]
```diff
@@ -132,6 +132,20 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
...
   }
 
+  /**
+   * Determines if the deposit can be updated, based on its current state.
+   *
+   * @param user ID
+   * @param id   the deposit ID
+   * @return
+   */
+  def canUpdate(user: String, id: UUID): Try[Unit] = {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: can this method be private?

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/StateInfo.scala (hunk 1 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/docs/StateInfo.scala]: In theory these errors could be caused by impatient users. For example delete a deposit on one browser tab while a submission on the same deposit is in progress on another. More likely by clients that don't look at the state. So the client should know these details. 

Issue 2016 is more about details like absolute file paths.

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 3]
```diff
@@ -218,6 +232,7 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
    */
   def writeDataMetadataToDeposit(dm: DatasetMetadata, user: String, id: UUID): Try[Unit] = for {
     deposit <- getDeposit(user, id)
+    _ <- canUpdate(user, id)
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Call inside `writeDatasetMetadataJson`, see https://github.com/DANS-KNAW/easy-deposit-api/pull/127#discussion_r270328540

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 4]
```diff
@@ -265,6 +280,7 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
   def writeDepositFile(is: => InputStream, user: String, id: UUID, path: Path): Try[Boolean] = {
     for {
       dataFiles <- getDataFiles(user, id)
+      _ <- canUpdate(user, id)
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: As above: call inside `dataFiles.write`. Take both log calls along.

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 5]
```diff
@@ -304,6 +320,7 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
    */
   def deleteDepositFile(user: String, id: UUID, path: Path): Try[Unit] = for {
     dataFiles <- getDataFiles(user, id)
+    _ <- canUpdate(user, id) //  deleting a file, is updating the deposit
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: As above: call inside `dataFiles.delete`.

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/StateInfo.scala (hunk 1 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/docs/StateInfo.scala]: @ 1) I suppose you mean some private `can(action: String, states: Seq[State])`

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala (hunk 4 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Sorry, got lost between App class and servlet.

**Pull request merged**